test/e2e/common/node/pod_admission.go

```diff
@@ -65,6 +71,71 @@ var _ = SIGDescribe("PodOSRejection", framework.WithNodeConformance(), func() {
 	})
 })
 
+var _ = SIGDescribe("PodRejectionStatus", func() {
+	f := framework.NewDefaultFramework("pod-rejection-status")
+	f.NamespacePodSecurityLevel = admissionapi.LevelBaseline
+	ginkgo.Context("Kubelet", func() {
+		ginkgo.It("should reject pod when the node didn't have enough resource", func(ctx context.Context) {
+			node, err := e2enode.GetRandomReadySchedulableNode(ctx, f.ClientSet)
+			framework.ExpectNoError(err, "Failed to get a ready schedulable node")
+
+			outOfCPUs := resource.MustParse("10")
+			outOfCPUs.Add(node.Status.Allocatable[v1.ResourceCPU])
+
+			// Create a pod that requests more CPU than the node has
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "pod-out-of-cpu",
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:  "pod-out-of-cpu",
+							Image: imageutils.GetPauseImageName(),
+							Resources: v1.ResourceRequirements{
+								Requests: v1.ResourceList{
+									v1.ResourceCPU: outOfCPUs, // requests more CPU than the node has
+								},
+							},
+						},
+					},
+					NodeName: node.Name, // Set the node to an node which has less CPU
+				},
+			}
+
+			pod = e2epod.NewPodClient(f).Create(ctx, pod)
+
+			// Check the pod status after the pod is rejected
+			var condition = func(oldPod *v1.Pod, newPod *v1.Pod) (bool, error) {
+				switch newPod.Status.Phase {
+				case v1.PodSucceeded:
+					return true, errors.New("pod succeeded unexpectedly")
+				case v1.PodFailed:
+					if strings.HasPrefix(newPod.Status.Reason, "OutOf") {
+						expectedStatus := oldPod.Status
+						// overwriting all fields that we expect are overridden by kubelet.
+						// All other fields must stay the same as before kubelet touched them
+						expectedStatus.Phase = newPod.Status.Phase
+						expectedStatus.Reason = newPod.Status.Reason
+						expectedStatus.Message = newPod.Status.Message
+						expectedStatus.StartTime = newPod.Status.StartTime
+						if !reflect.DeepEqual(expectedStatus, newPod.Status) {
+							return true, fmt.Errorf("unexpected status change: \nExpected:\n%s\n But got:\n%v", format.Object(expectedStatus, 1), format.Object(newPod.Status, 1))
+						}
+						return true, nil
+					} else {
+						return true, fmt.Errorf("pod failed with reason %s", newPod.Status.Reason)
+					}
+				}
+				return false, nil
+			}
+			err = e2epod.WaitForPodChange(ctx, f.ClientSet, pod.Namespace, pod.Name, f.Timeouts.PodStartShort, condition)
```

Comment:
I understand what you are trying to do. But our tests are often running on slow VMs. So there is a great potential for the race condition. Why not just make it straightforward - create Pod, get status, assign Node, check that it failed admission, compare statuses?

If you are worried that the Pod will get scheduled to the Windows node while we wait, you can set node name to some non-existing node name initially and then to the existing node name after saving it's original status

Reply:
>If you are worried that the Pod will get scheduled to the Windows node while we wait, you can set node name to some non-existing node name initially and then to the existing node name after saving it's original status

Once a pod is assigned to a node, it cannot be changed to another node. https://github.com/carlory/kubernetes/blob/master/pkg/registry/core/pod/storage/storage.go#L230

> Why not just make it straightforward - create Pod, get status, assign Node, check that it failed admission, compare statuses?

If a pod is created without a specified node, we can't get the status of the pod immediately, because the scheduler will change the status of the pod like below:

```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2024-10-19T08:42:54Z"
    message: '0/1 nodes are available: 1 Insufficient cpu. preemption: 0/1 nodes are
      available: 1 No preemption victims found for incoming pod.'
    reason: Unschedulable
    status: "False"
    type: PodScheduled
  phase: Pending
  qosClass: Burstable
```

I can poll the pod status until it is updated by the scheduler, save the status, and proceed to the next steps. However, there is a potential issue that we can not say that no other components, specifically custom controllers in some users' clusters, are changing the pod status. So, we can not guarantee that the saved status is the latest status observed by the kubelet when it is trying to reject the pod. 

I asked my teammate about the watch for a specific resource. He said that we can observe the pod's change sequence by watching the resource. Informer can leverage the watch mechanism to get old and new objects and pass them to the update event handler. 

I'm unsure whether we should consider other components or some unpredictable scenarios. The code is updated.